internal/cephfs/groupcontrollerserver.go

```diff
@@ -61,3 +73,601 @@ func (cs *ControllerServer) validateCreateVolumeGroupSnapshotRequest(
+		}
+		state := core.GetQuiesceState(data.State)
+		if state != core.Released {
+			return nil, status.Errorf(codes.Internal, "quiesce operation is in %s state", state)
+		}
+	}
+	var err error
+	defer func() {
+		if err != nil && !errors.Is(err, cerrors.ErrQuiesceInProgress) {
+			uErr := cs.deleteSnapshotsAndUndoReservation(ctx, vgs, cr, fsMap, req.GetSecrets())
+			if uErr != nil {
+				log.ErrorLog(ctx, "failed to delete snapshot and undo reservation: %v", uErr)
+			}
+		}
+	}()
+	snapshotResponses := make([]csi.CreateSnapshotResponse, 0)
+	for _, volID := range req.GetSourceVolumeIds() {
+		// Create the snapshot for the volumeID
+		clusterID := getClusterIDForVolumeID(fsMap, volID)
+		if clusterID == "" {
+			return nil, status.Errorf(codes.Internal, "failed to get clusterID for volumeID %s", volID)
+		}
+
+		req := formatCreateSnapshotRequest(volID, vgs.FsVolumeGroupSnapshotName,
+			clusterID,
+			req.GetSecrets())
+		var resp *csi.CreateSnapshotResponse
+		resp, err = cs.createSnapshotAndAddMapping(ctx, req, vg, vgs, cr)
+		if err != nil {
+			// Handle cleanup
+			log.ErrorLog(ctx, "failed to create snapshot: %v", err)
+
+			return nil, status.Errorf(codes.Internal,
+				"failed to create snapshot and add to volume group journal: %v",
+				err)
+		}
+		snapshotResponses = append(snapshotResponses, *resp)
+	}
+
+	response := &csi.CreateVolumeGroupSnapshotResponse{}
+	response.GroupSnapshot = &csi.VolumeGroupSnapshot{
+		GroupSnapshotId: vgs.VolumeGroupSnapshotID,
+		ReadyToUse:      true,
+		CreationTime:    timestamppb.New(time.Now()),
+	}
+
+	for _, r := range snapshotResponses {
+		r.Snapshot.GroupSnapshotId = vgs.VolumeGroupSnapshotID
+		response.GroupSnapshot.Snapshots = append(response.GroupSnapshot.Snapshots, r.Snapshot)
+	}
+
+	return response, nil
+}
+
+// createSnapshotAddToVolumeGroupJournal creates the snapshot and adds the
+// snapshotID and volumeID to the volume group journal omap. If the freeze is
+// true then it will freeze the subvolumes and subvolume groups before creating
+// the snapshot and unfreeze them after creating the snapshot. If the freeze is
+// false it will call createSnapshot and get the snapshot details for the
+// volume and add the snapshotID and volumeID to the volume group journal omap.
+// If any error occurs other than ErrInProgress it will delete the snapshots
+// and undo the reservation and return the error.
+func (cs *ControllerServer) createSnapshotAddToVolumeGroupJournal(
+	ctx context.Context,
+	req *csi.CreateVolumeGroupSnapshotRequest,
+	vgo *store.VolumeGroupOptions,
+	vgs *store.VolumeGroupSnapshotIdentifier,
+	cr *util.Credentials,
+	fsMap map[string]core.FSQuiesceClient) (
+	*[]csi.CreateSnapshotResponse,
+	error,
+) {
+	var err error
+	var resp *csi.CreateSnapshotResponse
+
+	responses := make([]csi.CreateSnapshotResponse, 0)
+	for _, volID := range req.GetSourceVolumeIds() {
+		err = fsQuiesceWithExpireTimeout(ctx, vgo.RequestName, fsMap)
```

Comment:
In lots of places where `fsMap` is passed, it might be nice to have functions like `fsMap.fsQuiesceWithExpireTimeout(...)`